src/voting_nodes.rs

```diff
@@ -0,0 +1,214 @@
+//! The `vote_stage` nodes on the `last_id` of the bank at a regular cadence
+
+use byteorder::{LittleEndian, ReadBytesExt};
+use counter::Counter;
+use crdt::{Crdt, NodeInfo};
+use hash::Hash;
+use log::Level;
+use signature::Pubkey;
+use std::collections::HashMap;
+use std::io::Cursor;
+use std::sync::atomic::AtomicUsize;
+use std::sync::{Arc, RwLock};
+use timing::timestamp;
+use transaction::Vote;
+
+pub struct VotingNodes {
+    pub nodes: HashMap<Pubkey, u64>,
+}
+
+fn make_debug_id(key: &Pubkey) -> u64 {
+    let buf: &[u8] = &key.as_ref();
+    let mut rdr = Cursor::new(&buf[..8]);
+    rdr.read_u64::<LittleEndian>()
+        .expect("rdr.read_u64 in fn debug_id")
+}
+
+impl VotingNodes {
+    pub fn new() -> Self {
+        VotingNodes {
+            nodes: HashMap::new(),
+        }
+    }
+
+    fn update_liveness(&mut self, debug_id: u64, id: Pubkey) {
+        //update the liveness table
+        let now = timestamp();
+        trace!(
+            "{:x} updating liveness {:x} to {}",
+            debug_id,
+            make_debug_id(&id),
+            now
+        );
+        *self.nodes.entry(id).or_insert(now) = now;
+    }
+
+    pub fn insert_vote(
+        &mut self,
+        crdt: &Arc<RwLock<Crdt>>,
+        pubkey: &Pubkey,
+        v: &Vote,
+        last_id: Hash,
+    ) {
+        let mut wcrdt = crdt.write().unwrap();
+
+        if wcrdt.table.get(pubkey).is_none() {
+            warn!(
+                "{:x}: VOTE for unknown id: {:x}",
+                wcrdt.debug_id(),
+                make_debug_id(pubkey)
+            );
+            return;
+        }
+        if v.contact_info_version > wcrdt.table[pubkey].contact_info.version {
```

Comment:
i am starting to think that the contact_info is kind of pointless.  you can use a different keypair when your ip or port changes.

Reply:
Should we deprecate it? May be a separate ticket?